Adding num-util crate with FixedDecimal

sffc opened the pull request:
This is an MVP of the FixedDecimal type.

In a second PR, I would like to add a trait like `PluralOperandProducer` with functions for the ~6 plural operands (n, i, f, d, etc), and FixedDecimal can implement that trait.

In a third PR, I would like to add some additional functions necessary for number formatting, such as the ability to round the number.

zbraniecki commented on components/num-util/src/digit_converter.rs:
```diff
@@ -0,0 +1,101 @@
+        debug_assert_le!(d, 9);
+        let cp: u32 = self.zero_char as u32 + d as u32;
+        // TODO: Is there an unchecked version we can use here, if we check zero_char earlier?
+        cp.try_into().unwrap_or('\u{FFFD}')
```
Could we use `d - b'0'`?

zbraniecki commented on components/num-util/src/digit_converter.rs:
```diff
@@ -0,0 +1,101 @@
+            }
+            None => None,
+        }
+    }
```
how is it different from:

```rust
let b = c as u8;
if b >= b'0' && b <= b'9' {
    Some(c)
} else {
    None
}
```
?

sffc replied:
I think that doesn't work with non-ASCII code points?

There's a unit test below that shows how this works with other numbering systems.

Do you think we should use a different code path for ASCII digits for better efficiency?

sffc replied:
OK, I thought about it and right now there is only 1 call site for this class, and it can be replaced by a simple `b'0' + d`.  I removed DigitConverter from this PR and will add it back again in the future if it is needed.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+            value.abs() as usize
+        };
+        let mut result = FixedDecimal::from_ascending(UintIterator::from(value_abs)).unwrap();
+        result.is_negative = value < 0;
```
this will set `is_negative = false` for `-0`.

sffc replied:
Good point, but can `-0` be represented by an `isize`?

zbraniecki replied:
good point :)

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
````diff
@@ -0,0 +1,517 @@
+    /// assert_eq!(9, dec.digit_at(2));
+    /// assert_eq!(0, dec.digit_at(3));
+    /// ```
+    pub fn digit_at(&self, magnitude: i16) -> u8 {
````
Why doesn't it return an `Option`?

sffc replied:
There are 5 "zones" to consider:

1. magnitude is smaller than lowest_magnitude
2. magnitude is larger than lowest_magnitude but smaller than the smallest nonzero digit
3. magnitude is a nonzero digit
4. magnitude is smaller than upper_magnitude but larger than the largest nonzero digit
5. magnitude is larger than upper_magnitude

Zones 2, 3, and 4 are the visible digits declared by the FixedDecimal settings.  Zones 1 and 5 are out of bounds and could be considered None.  However, I do not see this as a useful distinction to make.  First, zero is a sensible and correct fallback.  Second, if the user wants to know the bounds, they should use the `magnitude_range` function.  I therefore leaned on the side of making this function as efficient as possible.

zbraniecki replied:
It feels fishy to me. I never encountered an API in Rust that would return `0` or some default value from a call that asks for an item at a position that is not available.
@Manishearth - how does it sound to you?

sffc replied:
It's not that the digit is "not available" (as stated above, ones 2 and 4 are definitely expected to fall back to 0); it's just that if the digit did exist, it would be a 0.  If you have the number 123, what is the digit in the thousands place?  In the tenths place?  They are both 0, even though they are out of range of the number 123.

zbraniecki replied:
I see. Yes, that makes sense. Thank you for explaining this.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+            // The following line can't fail: magnitude <= self.magnitude, by
+            // the if statement above, and u16::MAX == i16::MAX - i16::MIN, and
+            // usize is asserted to be at least as big as u16.
+            let j: usize = (self.magnitude as i32 - magnitude as i32) as usize;
```
you don't need to provide `usize` type twice.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+    }
+}
+
+// TODO(review): This should be fmt::Display, but that is very no_std-unfriendly for code size.
```
can you elaborate? Why wouldn't we want `Display` here? If it's not used, it'll be shaken off.

sffc replied:
Library code might want to use ToString, but if you implement Display, then ToString uses the heavier Display code path.

zbraniecki replied:
Can you improve the comment then? Maybe:

```
`Display` provides a heavier implementation of `ToString` while we want to explicitly handle `ToString` for library use.
```

Also, can we implement both to allow `FixedDecimal` to be used in alloc-free way?

sffc replied:
`TODO(review)` means that I wanted you to evaluate a decision I made but that I wasn't 100% confident in.  First, do you agree with me that it is a worthwhile win for code size to separate ToString from Display?  If so, then yes, I can look into also implementing Display.

zbraniecki replied:
Yep, agree.

sffc replied:
If I do `impl Display for FixedDecimal`, then the `impl ToString for FixedDecimal` fails to compile with E0119, since the standard library has a generic ToString for types implementing Display.

sffc replied:
I went ahead and did `impl Display for FixedDecimal`, and supplemented it with new methods `write_to` and `write_len` if one needs a more efficient implementation.

zbraniecki commented:
> In a second PR, I would like to add a trait like PluralOperandProducer with functions for the ~6 plural operands (n, i, f, d, etc), and FixedDecimal can implement that trait.

I'm wondering if there's a value in having `FixedDecimal` implement `PluralOperandProducer` (in Rust ling it'll be `PluralOperandsType`), or would it be simpler to just have `From<FixedDecimal> for PluralOperands`.

My reasoning is that any customer of `PluralOperands` will use the operands as immutable, care only about the operands and need them available. Having methods that return those values from the underlying data stored by `FixedDecimal` will have a performance cost, and the benefit of having the underlying struct be a `FixedDecimal` that user can mutate is diminished.

So maybe just being able to produce one from another is enough?

sffc commented:
`From<FixedDecimal> for PluralOperands` could work. The immutability is a plus. A downside is that it requires pre-calculating all the operands, even if the specific ruleset doesn't require them.

sffc commented:
I added a benchmark file.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+            is_negative: false,
+        }
+    }
+}
```
How common do you expect someone to want to create a `FixedDecimal` of value `0`?

I imagine that mostly being useful if there's a struct that has a field of `FixedDecimal` type and we'd want to have `Default` derived for it.
But that feels weird - after all, if someone wants a `FixedDecimal` they're most likely gonna accept some input and parse it into `FixedDecimal`, no? What would a default value of `0` give, if they still need/want to later take that input and reassign the field?

sffc replied:
Not sure what you're suggesting here.  Are you suggesting to not implement the `Default` trait because you question whether it will be widely used?

I implemented the trait because it seems useful for a data type to have a Default trait as its only constructor.  Also, the `from` methods call `Default` before appending things to the digit vector.  Without this trait, I would probably still have a method returning the same thing, but I could make it private.

zbraniecki replied:
I'm trying to understand the what `FixedDecimal` is, and it seems to me a particular type of `Number` with special methods.
You wouldn't have `Number:;default()` method, right? What would it mean? If it existed, then it'd likely return `0` for `Default`, but it most likely would not exist.

That's why I'm questioning whether we should have `Default` trait on it at all.

As to what you're calling in `from`, you can just construct it manually.

sffc replied:
I think 0 is a very sensible default for FixedDecimal: it's a decimal number with no significant digits.

sffc replied:
> I'm trying to understand the what FixedDecimal is

Here are some attempts at answering this question.  Hopefully one of them sticks:

1. A FixedDecimal is a base-10 number in which every position (power of 10) is mapped to a decimal digit between 0 and 9.
2. A FixedDecimal is a base-10 number with zero or more significant (nonzero) digits.
3. A FixedDecimal is a numbering-system-agnostic representation of a base 10 number as humans would think about it.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+            // Take only up to std::i16::MAX values so that we have enough capacity
+            if x > std::i16::MAX as usize {
+                return Err(Error::Limit);
+            }
```
`digits_iter.take(std::i16::MAX).enumerate()`

zbraniecki replied:
and then you can check if the iterator is done at the end and throw if its not.

You can also use `ExactSizeIterator` which is reasonable I think in this case - I don't expect anyone to throw streaming lexer at this and try to parse digit by digit an unknown length source.

sffc replied:
> and then you can check if the iterator is done at the end and throw if its not.

How do you do that?  `.take()` moves the iterator so I can't call any methods on it later, right?

> You can also use ExactSizeIterator which is reasonable I think in this case - I don't expect anyone to throw streaming lexer at this and try to parse digit by digit an unknown length source.

Actually no, because the common case is that you're pulling digits off an integer.  To determine how many digits are in the integer, you would need to either perform an expensive `log10` on the input, or run the number through a series of if/else statements checking the number against different powers of 10.  The benefit of an ExactSizeIterator doesn't seem to outweigh the cost.

zbraniecki replied:
> How do you do that? .take() moves the iterator so I can't call any methods on it later, right?

uh, right! Sorry for noise. I was confused about the role of `take`.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+                    Some(v) => mem[v] = d,
+                    // This error should be obsolete after X is made generic
+                    None => return Err(Error::Limit),
+                }
```
```rust
mem[v] = X.checked_sub(i).ok_or(Error::Limit)?;
```

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+            self.lower_magnitude = match self.lower_magnitude.checked_add(delta) {
+                Some(v) => v,
+                None => return Err(Error::Limit),
+            };
```
In both cases you should be able to use `ok_or` and `?`, no?

zbraniecki commented on components/num-util/src/uint_iterator.rs:
```diff
@@ -0,0 +1,59 @@
+use num_integer::Integer;
+use num_traits::{FromPrimitive, ToPrimitive, Unsigned};
```
could we skip it for now and just support primitive types?

sffc replied:
I'm not opposed to that.  However, we should do so knowing the functionality we lose:

- No support for `num::BigUint`.
- No support for arbitrary-precision numbers (limited by an `isize`), unless we want to consider other ways of achieving this, such as `FromStr`.  (I would rather leave`FromStr` to a later ticket, as it is somewhat involved.)

sffc commented on components/num-util/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+num-traits = "0.2.12"
+num-integer = "0.1.43"
+static_assertions = "1.1.0"
+more-asserts = "0.2.1"
```
By "unstable" do you mean those with major version 0?

- `num-traits` and `num-integer`, which are in the meta-crate `num`, are my attempt at generalizing the `from` methods for FixedDecimal.  If we remove them, you will still be able to construct a FixedDecimal from hard-coded types like `isize`, but not from `BigUint` or other types.  My guess is that some of the `num` crate might end up in the stdlib one day, but I do not know their plans.  I have no problem removing these if it's better without them.
- `more-asserts` was so I could get `>` and `<` asserts.  I think removing this one would make code quality go down, and it doesn't affect what we ship since these are only used in debug mode, so I vote to keep it.

zbraniecki replied:
I see, thank you.

> num-traits and num-integer, which are in the meta-crate num, are my attempt at generalizing the from methods for FixedDecimal. If we remove them, you will still be able to construct a FixedDecimal from hard-coded types like isize, but not from BigUint or other types.

My first feeling is that it's a bit premature to aim for compatibility with a custom type like `biguint`.

I'd like to suggest that we first write this for just primitive types, potentially using the model such as https://github.com/zbraniecki/pluralrules/blob/master/intl_pluralrules/src/operands.rs#L183-L186 and then we evaluate scaling it up to handle non-standard types such as `biguint` or some other version of it if another library gain traction.

sffc commented on components/num-util/benches/isize.rs:
```diff
@@ -0,0 +1,43 @@
+    dist.sample_iter(rng)
+        .take(1000)
+        .map(|v| v as isize)
+        .collect()
```
I specifically chose a PRNG with a hard-coded seed to make the random numbers the same each time the benchmark is run.

sffc commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,517 @@
+    /// - Must not include leading or trailing zeros
+    /// - Length must not exceed (magnitude - lower_magnitude + 1)
+    // TODO: Consider using a nibble array
+    digits: SmallVec<[u8; 8]>,
```
It seems "obvious" to me that `SmallVec` is the right tool for the job compared to `Vec` and `Box<[u8]>`.  Do you have experience that suggests that `SmallVec` *isn't* always the objectively best option for vectors with a small number of bytes on the stack?  I personally haven't seen or read anything that has given me pause when picking SmallVec for a short array of bytes.

To answer your questions specifically:

1. I have previously compared SmallVec to Vec (actually SmallStr to String), and I found negligible difference in code size, with SmallStr coming out slightly smaller than String.  I haven't looked at speed or memory, only code size, and I haven't done it for this specific use case.  For optimal velocity, I would rather compare things like SmallVec vs Vec in a vacuum and make general recommendations.
2. Mutable operations will likely be very common on this class, specifically when we implement rounding behavior.

zbraniecki replied:
> It seems "obvious" to me that SmallVec is the right tool for the job compared to Vec and Box<[u8]>. Do you have experience that suggests that SmallVec isn't always the objectively best option for vectors with a small number of bytes on the stack?

Yes. In several scenarios `SmallVec` was slower than `Vec`.
Now, I understand that `SmallVec` gives you stack, but I'd test it yourself.

> Mutable operations will likely be very common on this class, specifically when we implement rounding behavior.

I hear you, so mutability will be common. Can you give me example of use case where one would allocate a `FixedDecimal` and then manipulate it?
I'm ok with imaginary based on your experience. Would it be more of a hot-code in other ICU methods, or userland code that would use it?

sffc replied:
1. OK, since I have a benchmark file now, it should be easy to do a quick speed check.  I'll get to that on Monday or Tuesday.
2. I anticipate that users will generally create a FixedDecimal and give it (by value) to NumberFormat, which will then perform mutative operations like rounding to the nearest cent or scaling and rounding according to compact notation preferences.  I do not anticipate that userland code would generally mutate a FixedDecimal.

I went back and forth multiple times on whether to traitify FixedDecimal rather than giving a concrete struct.  I acknowledge that a trait seems to make more sense, but I settled on a struct.  I did not spend the time to document why I did this.  I can rehash that decision if you dispute it.

zbraniecki replied:
Not disputing, I'm all in favor of starting with a concrete type even for things we may later add a trait for.

I'm trying to understand the use.
From what you're saying the typical use will be to use `From` from some source value to a `FixedDecimal` that then will be passed to some `NumberFormat` related code that will adjust the rounding and adjusting the visible digits.

That convinces me that mutability is a hot code for this, and that `Box<[u8]>` is not the right use case.

zbraniecki replied:
I did some initial benchmarks and `SmallVec` does seem to be the most efficient!

sffc commented:
Okay, I pushed my changes:

- Removed the dependency on the unstable `num` crate
- Added a sign field to UintIterator, renaming it to IntIterator
- Used `macro_rules!` to implement FixedDecimal and IntIterator for all built-in integer types
- Implemented other code review feedback like `.ok_or()`

sffc commented on components/num-util/src/uint_iterator.rs:
```diff
@@ -0,0 +1,102 @@
+impl_from_unsigned_integer_type!(u16);
+impl_from_unsigned_integer_type!(u8);
+
+impl<T> Iterator for IntIterator<T>
```
@zbraniecki Should I stick this part into the `macro_rules!` as well, instead of using all these traits?

sffc replied:
I decided to do it.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,470 @@
+            match self.digits.get(j) {
+                Some(v) => *v,
+                None => 0, // Trailing zero
+            }
```
consider `*self.digits.get(j).unwrap_or(&0)`

sffc replied:
I prefer the way I had written it because it makes it clearer that the `0` refers to a trailing zero.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
````diff
@@ -0,0 +1,470 @@
+    /// assert_eq!(0..=2, dec.magnitude_range());
+    /// ```
+    pub fn magnitude_range(&self) -> RangeInclusive<i16> {
+        RangeInclusive::new(self.lower_magnitude, self.upper_magnitude)
````
consider `self.lower_magnitude..=self.upper_magnitude`

sffc commented on components/num-util/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[dependencies]
+smallvec = "1.4.1"
+static_assertions = "1.1.0"
+more-asserts = "0.2.1"
```
I prefer being explicit about assumptions in code, even benign ones like "usize is at least as big as u16".  I would like to keep `static_assertions` for this purpose.  This is an assertion for the compiler: ICU4X should refuse to compile if that condition is not met.  It blows my mind that static assertions aren't in the standard library like they are in C++, where I use them frequently.  If static assertions ever make it there, it will be an easy change to make without needing to understand the intricacies of FixedDecimal.

I don't feel as strongly about `more-asserts`.  I added them because the error message is nicer if the assertions were to ever fail.  But, that's probably not necessary, because these assertions shouldn't ever fail.  I'll remove `more-asserts`.

The assertion crates are big winners in the size/benefit tradeoff, since they are not intended to add any runtime bloat.  I hear what you're saying about compile time though.  I haven't measured the degree to which that is affected.

sffc commented on components/num-util/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+rand = "0.7.3"
+criterion = "0.3"
+rand_pcg = "0.2.1"
+rand_distr = "0.2.2"
```
OK.  On the other hand, at least with runtime dependencies like `smallvec`, maybe it's better to enforce the minor and patch versions, because code size can and will change between `smallvec` minor and patch versions, which can and will create inconsistencies when people run those tests.

I changed the dev dependencies as suggested, but left the runtime dependencies with patch versions.

Follow-up: https://github.com/unicode-org/icu4x/pull/141

sffc replied:
Changed to major.minor based on conclusion in #175

sffc commented on components/num-util/benches/isize.rs:
```diff
@@ -0,0 +1,43 @@
+
+    // Note: this could be bench_function_with_inputs, but there are 1000 random inputs.
+    // Instead, consider all inputs together in the same benchmark.
+    c.bench_function("isize/larger", |b| {
```
I tried replacing `nums` with `black_box(nums)`, but that gives

```
cannot move out of `nums`, a captured variable in an `FnMut` closure

move occurs because `nums` has type `std::vec::Vec<isize>`, which does not implement the `Copy` traitrustc(E0507)
```

As the code comment discusses, I initially tried using Benchmarking With Inputs, but decided against it because I thought it was more valuable to get an overall picture of the performance over a distribution of input numbers rather than one individual input number at a time.

zbraniecki replied:
I think you'd do `b.iter(|| nums.iter().map(|v| black_box(*v)).map(FixedDecimal::from).count());`

but I don't think at the moment LLVM is doing constant folding in this case, so it's up to you.

sffc commented:
@elango: the clippy check is failing on a file that no longer exists.  Can you take a look?

sffc commented:
@echeran: the clippy check is failing on a file that no longer exists.  Can you take a look?

echeran commented:
> @echeran: the clippy check is failing on a file that no longer exists. Can you take a look?

I create PR #179 to fix the issue.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
````diff
@@ -0,0 +1,538 @@
+    /// let dec = FixedDecimal::from(42).multiplied_pow10(3).expect("Bounds are small");
+    /// assert_eq!("42000", dec.to_string());
+    /// ```
+    pub fn multiplied_pow10(mut self, delta: i16) -> Result<Self, Error> {
````
hmm, one of the main values of having this variant of the method is that you could use it on an immutable `FixedDecimal` to produce a new one.
The way you use it here requires a mutable reference which diminishes the value of building a new one.

There are probably smart optimizations we can add later, but for now, maybe just make it immutable and `self.clone().multiply_pow10(delta)` ?

sffc replied:
The argument is `mut self`, not `&mut self`.  I think it is equivalent to an argument with type `self` with a line in the function such as `mut retval = self;`.  Is it confusing/bad style for an argument to be `mut self` but be by value and not reference?

zbraniecki replied:
Ah, I misread the code. It's all good if you need it mutable inside.

zbraniecki commented on components/num-util/src/fixed_decimal.rs:
```diff
@@ -0,0 +1,538 @@
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let dec = FixedDecimal::from(-5000).multiplied_pow10(-2).expect("Bounds are small");
+    /// let mut result = String::new();
```
Wouldn't you want to `String::with_capacity(dec.write_len())` ?

zbraniecki commented on components/num-util/src/lib.rs:
```diff
@@ -0,0 +1,4 @@
+pub mod fixed_decimal;
```
It would be nice to have some basic crate-level intro to what `FixedDecimal` is and a simple example of canonical use.

sffc commented:
I added crate-level docs and made the change you suggested with the `String::with_capacity` example.  I also moved `Error` to `lib.rs`, to be more consistent with UnicodeSet and how we plan to approach errors on the component level.  I will merge this PR this afternoon unless anyone has more feedback.

sffc commented:
(I still need a rubber stamp since reviews are dismissed upon subsequent commits)

zbraniecki commented:
> (I still need a rubber stamp since reviews are dismissed upon subsequent commits)

I'd like to fix that. If I r+'ed it, it's good to go.
